### ISSUE TITLE
Improve file watcher handling performance

### ISSUE THREAD

**Issue by lptr** [gradle/gradle#11888]
File watchers currently add some unnecessary overhead in several places:

1) for all OSs: we use `Files.isDirectory()` to check if the parent directory to watch is indeed an (existing) directory here: https://github.com/gradle/gradle/blob/3f9b7efefeddafdfb8dfd048d342011d34321abb/subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/DefaultWatchingVirtualFileSystem.java#L82-L85
    
    This is a slow operation, and it amounts to 130 ms of processing for ~1800 watched directories on macOS.
    
    We could fix this easily for Windows, where in the end we only care about the drive letters to watch.
    
    We could also get rid of most of the overhead on macOS (provided we continue using hierarchical `FSEvents`) by ignoring directories whose ancestors are already watched earlier.
    
    On Linux we would probably need something more elaborate, reusing the VFS itself to find the first existing parent candidate. We'd need some handling for when the directory has been removed in the meantime. (The latter is a possibility with other OSs as well, but the window for the race condition is smaller for them.)

2) we canonicalize the paths we watch explicitly. This shouldn't be necessary, as the OS should determine the canonical path on its own. We will most probably switch (back) to using canonicalized paths in the VFS later anyway (to let the VFS work with symlinks). This seems to be responsible for a smaller part of the overhead.

See async profiler results for macOS: [onlyVfsRetention.zip](https://github.com/gradle/gradle/files/4059160/onlyVfsRetention.zip)

---
cc: @gradle/build-cache

**Comment by lptr** [comment]
For 1) we also don't need to check if the parent is a directory unless the item we are checking is a missing file.

**Comment by lptr** [comment]
We could also drop any missing files from the VFS at the end of the build.

**Comment by lptr** [comment]
If we don't check for the existence of parent directories we know should be there, we are down to around 45-50 ms (from 130) for 1). Step 2) is still at 28 ms.

**Comment by lptr** [comment]
Another potential improvement is to change the API to `native-platform` to avoid conversion of the watched paths. Right now we do `String` -> `Path` -> `String` -> `File` -> `File.canonicalFile` -> `String` with each watched path.

If we removed the canonicalization on the native side, we could either use `String` or `Path`, though using `Path` make the feature unavailable on Java 5 and 6, which is still supported by `native-platform`.

**Comment by lptr** [comment]
Here's another async profiler result, after https://github.com/gradle/gradle/pull/11890/commits/c84814f902ccc0b3bd381d34824821c67655ed81: [onlyVfsRetention-after-directory-lookup.zip](https://github.com/gradle/gradle/files/4059351/onlyVfsRetention.zip)

**Comment by lptr** [comment]
Here's how I took the Async profiler measurements using https://github.com/gradle/santa-tracker-performance:

```text
$ gradle-profiler/build/install/gradle-profiler/bin/gradle-profiler \
    --project-dir santa-tracker-java \
    --scenario-file santa-tracker-java.scenarios \
    --gradle-version "$GRADLE_VERSION" \
    --warmups 2 \
    --iterations 10 \
    --profile async-profiler \
    onlyVfsRetention
```

**Comment by lptr** [comment]
We are getting closer to the results without the watcher registration overhead. Here's where we are on Linux:

Results *without* the optimizations in https://github.com/gradle/gradle/commit/c84814f902ccc0b3bd381d34824821c67655ed81:

![chart (16)](https://user-images.githubusercontent.com/495366/72349818-87aaa000-36dd-11ea-97fd-0d31a3b862cd.png)

Results *with* the optimizations:

![chart (15)](https://user-images.githubusercontent.com/495366/72349703-503bf380-36dd-11ea-8810-5247d0bac940.png)

Results *with no registration overhead* from end of November (i.e. our ideal goal):

![chart (17)](https://user-images.githubusercontent.com/495366/72349874-a315ab00-36dd-11ea-91be-1d3d4ce3c885.png)

**Comment by lptr** [comment]
I've ran async profiler on the Linux variant which shows the same behavior for registering the watchers (even though it needs to register a lot more than macOS):

```text
Spent 29 ms figuring out what to watch
Watching 1836 directories to track changes between builds
Spent 47 ms watching 1836 directories for file system events
```

...but it takes a lot longer to actually read the changes:

```text
Virtual file system retention is an incubating feature.
Marking as changed since last build: /home/lptr/Workspace/gradle/santa-tracker-performance/santa-tracker-java/village/src/main/java/com/google/android/apps/santatracker/village/SnowFlake.java
Received 1 file system events since last build
Spent 14 ms processing file system events since last build
```

I'm surprised that `stopWatching()` doesn't show up in the stacks despite taking a comparable amount of time.

See profiler output: 
[linux-with-no-directory-check.zip](https://github.com/gradle/gradle/files/4059549/linux-with-no-directory-check.zip)

**Comment by wolfs** [comment]
Here are some ideas:
```
I see a bunch of ways to decrease the time spent registering the watchers:
– (for macOS with FSEvents) register the project root to decrease the number of watchers needed to register
- register watchers in parallel
- register watchers after the build has returned control to the user
- do not unregister watchers between builds
- register watchers while the build is running, immediately when a location is touched
```

**Comment by lptr** [comment]
I think we should focus on keeping the watchers alive during builds instead of registering them in parallel/after the build has returned control to the user. We will need the watchers to be alive during the build anyway for some of the features we are planning.

We could split this up into two phases if necessary:

1) update watcher registrations at the end of every build
2) update watcher registrations whenever the VFS is touched (i.e. watch any newly discovered location; and stop watching locations that are invalidated/deleted).

**Comment by lptr** [comment]
I've added #11901 and #11902.

**Comment by wolfs** [comment]
Let's close this now that the first round of improvements has been done.
